core/src/main/java/io/grpc/ServerInterceptors.java

```diff
@@ -179,101 +173,96 @@ public InputStream parse(final InputStream stream) {
-    public static <ReqT, RespT> InterceptCallHandler<ReqT, RespT> create(
-        ServerInterceptor interceptor, ServerCallHandler<ReqT, RespT> callHandler) {
-      return new InterceptCallHandler<ReqT, RespT>(interceptor, callHandler);
+  private static class InterceptCallHandler implements ServerCallHandler {
+
+    public static <ReqT, RespT> InterceptCallHandler create(
+        ServerInterceptor interceptor, ServerCallHandler callHandler) {
+      return new InterceptCallHandler(interceptor, callHandler);
     }
 
     private final ServerInterceptor interceptor;
-    private final ServerCallHandler<ReqT, RespT> callHandler;
+    private final ServerCallHandler callHandler;
 
     private InterceptCallHandler(ServerInterceptor interceptor,
-                                 ServerCallHandler<ReqT, RespT> callHandler) {
+                                 ServerCallHandler callHandler) {
       this.interceptor = Preconditions.checkNotNull(interceptor, "interceptor");
       this.callHandler = callHandler;
     }
 
     @Override
-    public ServerCall.Listener<ReqT> startCall(
+    public <ReqT, RespT> ServerCall.Listener<ReqT> startCall(
         MethodDescriptor<ReqT, RespT> method,
         ServerCall<RespT> call,
         Metadata headers) {
       return interceptor.interceptCall(method, call, headers, callHandler);
     }
   }
 
-  private static <OReqT, ORespT, WReqT, WRespT> ServerMethodDefinition<WReqT, WRespT> wrapMethod(
-      final ServerMethodDefinition<OReqT, ORespT> definition,
-      final MethodDescriptor<WReqT, WRespT> wrappedMethod) {
-    final ServerCallHandler<WReqT, WRespT> wrappedHandler = wrapHandler(
-        definition.getServerCallHandler(),
-        definition.getMethodDescriptor(),
-        wrappedMethod);
-    return ServerMethodDefinition.create(wrappedMethod, wrappedHandler);
-  }
-
-  private static <OReqT, ORespT, WReqT, WRespT> ServerCallHandler<WReqT, WRespT> wrapHandler(
-      final ServerCallHandler<OReqT, ORespT> originalHandler,
-      final MethodDescriptor<OReqT, ORespT> originalMethod,
-      final MethodDescriptor<WReqT, WRespT> wrappedMethod) {
-    return new ServerCallHandler<WReqT, WRespT>() {
+  private static ServerCallHandler wrapHandler(
+      final ServiceDescriptor originalDescriptor,
+      final ServerCallHandler originalHandler) {
+    return new ServerCallHandler() {
       @Override
-      public ServerCall.Listener<WReqT> startCall(
-          final MethodDescriptor<WReqT, WRespT> method,
-          final ServerCall<WRespT> call,
+      public <ReqT, RespT> ServerCall.Listener<ReqT> startCall(
+          final MethodDescriptor<ReqT, RespT> method,
+          final ServerCall<RespT> call,
           final Metadata headers) {
-        final ServerCall<ORespT> unwrappedCall = new PartialForwardingServerCall<ORespT>() {
+        final MethodDescriptor originalMethod =
+            originalDescriptor.getMethod(method.getFullMethodName());
```

Comment:
So this is the line that shows the type of tradeoff we are making by having `MethodDescriptor` contain the int index, instead of some opaque object. Seems okay in this case, but I may try to think about other cases that would be impacted.